torch/jit/__init__.py

```diff
@@ -2110,6 +2116,68 @@ def _get_script_class(name):
                            "Did you forget to import it?".format(name))
     return _script_classes[name]
 
+# qualified_name => list[(overload decl, overload default args)]
+_overloaded_fns = {}
+
+# qualified name => list[compiled fns]
+_compiled_overloaded_fns = {}
+
+def _overload(func):
+    qual_name = _qualified_name(func)
+    global _overloaded_fns
+    fn_overload_list = _overloaded_fns.get(qual_name)
+    if fn_overload_list is None:
+        fn_overload_list = []
+        _overloaded_fns[qual_name] = fn_overload_list
+    signature = torch.jit.annotations.get_signature(func)
+    if signature is None:
+        raise RuntimeError("Must explicitly add type annotations to overloaded functions: {obj}").format(func)
+    fn_overload_list.append((torch.jit.get_jit_def(func).decl(), get_default_args(func)))
+    return func
+
+def _compile_function_with_overload(qual_name, impl_fn, overload_decl, overload_defaults):
+    impl_ast = torch.jit.get_jit_def(impl_fn)
+    _frames_up = 0
+    _rcb = _gen_rcb(impl_fn, _frames_up)
+    fn = torch._C._jit_script_compile_overload(qual_name, overload_decl, impl_ast, _rcb, overload_defaults)
+    return fn
+
+def _get_overloads(obj):
+    # check for cached compiled fns
+    qual_name = _qualified_name(obj)
+    global _compiled_overloaded_fns
+    compiled_overloads = _compiled_overloaded_fns.get(qual_name, None)
+    if compiled_overloads is not None:
```

Comment:
This doesn't work if we call `_get_overloads()` then add another overload after. The second overload will not ever get compiled. Do we need a cache here? We don't have anything similar for other recursively compiled functions.

Reply:
What is the use case you are imagining? This is still an internal only api so I don't think it applies but something to think about if we make it public. We don't strictly need a cache it's true.

Reply:
If we add a new overload after using the overload in the scripted function, it will not be registered. Let's remove the cache, or make the existing caching logic correct. Even if it's unlikely I think it'll be very surprising if we ever run into this.